homeassistant/components/zimi/light.py

```diff
@@ -0,0 +1,136 @@
+
+    def __init__(self, light: ControlPointDevice, api: ControlPoint) -> None:
+        """Initialize an ZimiLight."""
+
+        self._attr_unique_id = light.identifier
+        self._attr_should_poll = False
+        self._light = light
+        self._light.subscribe(self)
+        self._state = False
+        self._brightness = None
+        if self._light.type == "dimmer":
+            self._attr_color_mode = ColorMode.BRIGHTNESS
+            self._attr_supported_color_modes = {ColorMode.BRIGHTNESS}
+        else:
+            self._attr_color_mode = ColorMode.ONOFF
+            self._attr_supported_color_modes = {ColorMode.ONOFF}
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, light.identifier)},
+            name=self._light.name,
+            suggested_area=self._light.room,
+            via_device=(DOMAIN, api.host + ":" + str(api.port)),
+        )
+        self.update()
+        _LOGGER.debug("Initialising %s in %s", self.name, self._light.room)
+
+    def __del__(self):
+        """Cleanup ZimiLight with removal of notification."""
+        self._light.unsubscribe(self)
+
+    @property
+    def name(self) -> str:
+        """Return the display name of this light."""
+        return self._name.strip()
+
+    @property
+    def available(self) -> bool:
+        """Return True if Home Assistant is able to read the state and control the underlying device."""
+        return self._light.is_connected
+
+    @property
+    def brightness(self) -> int | None:
+        """Return the brightness of the light.
+
+        This method is optional. Removing it indicates to Home Assistant
+        that brightness is not supported for this light.
+        """
+        return self._brightness
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if light is on."""
+        return self._state
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Instruct the light to turn on.
+
+        You can skip the brightness part if your light does not support
+        brightness control.
+        """
+
+        _LOGGER.debug(
+            "Sending turn_on(brightness=%d) for %s",
+            kwargs.get(ATTR_BRIGHTNESS, 255) * 100 / 255,
+            self.name,
+        )
+
+        if self._light.type == "dimmer":
+            await self._light.set_brightness(
+                kwargs.get(ATTR_BRIGHTNESS, 255) * 100 / 255
+            )
+        else:
+            await self._light.turn_on()
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Instruct the light to turn off."""
+
+        _LOGGER.debug("Sending turn_off() for %s", self.name)
+
+        await self._light.turn_off()
+
+    def notify(self, _observable):
+        """Receive notification from light device that state has changed."""
+
+        _LOGGER.debug("Received notification() for %s", self.name)
+        self.schedule_update_ha_state(force_refresh=True)
+
+    def update(self) -> None:
```

Comment:
better to use async_update, but tbh I doubt you actually need this. If I understand you correctly the controller will push the data, so better to update `self._light` in the callback and do an `self._async_write_ha_state` if needed.

Reply:
The controller just pushes a notification that a change has occurred, which I then read and use to set the data in update().   For some reason I thought I needed to do the schedule_update_ha_state() but if not I will change as you suggest.

Reply:
if you're outside the event loop `schedule_update_ha_state` is the correct one.